src/main/java/net/quarrymod/blockentity/machine/tier3/QuarryBlockEntity.java

```diff
@@ -181,98 +198,111 @@ private void refreshUpgrades() {
         fortuneLevel = 0;
         isSilkTouch = false;
 
-        quarryUpgradesSlotGroup.executeForAll((stack) -> {
-            if (!stack.isEmpty() && stack.getItem() instanceof IQuarryUpgrade) {
-                ((IQuarryUpgrade) stack.getItem()).process(this, stack);
+        quarryUpgradesSlotGroup.executeForAll(stack -> {
+            if (!stack.isEmpty() && stack.getItem() instanceof IQuarryUpgrade upgradeItem) {
+                upgradeItem.process(this, stack);
             }
         });
     }
 
+    @SuppressWarnings("ConstantConditions")
     private void tickQuarryLogic() {
-        if (excavationState != ExcavationState.Complete) {
-            if (miningSpentedEnergy < getEnergyPerExcavation()) {
-                final long euNeeded = getEnergyPerExcavation() / getTiksPerExcavation();
-                final long euAvailable = Math.min(euNeeded, getStored());
-                if (euAvailable > 0d) {
-                    useEnergy(euAvailable);
-                    miningSpentedEnergy += euAvailable;
-                    setExcavationState(ExcavationState.InProgress);
-                } else {
-                    setExcavationState(ExcavationState.NoEnergyIncome);
-                }
-            }
+        if (excavationState == ExcavationState.Complete) {
+            return;
+        }
+        if (miningEnergySpend < getEnergyPerExcavation()) {
+            performEnergyReduction();
+        }
 
-            if (miningSpentedEnergy >= getEnergyPerExcavation()) {
-                if (excavationWorkType == ExcavationWorkType.ExtractTube) {
-                    tryDrillOutTube();
-                } else {
-                    if (targetOrePos == null && (excavationState == ExcavationState.InProgress
-                        || excavationState == ExcavationState.NoOreInCurrentPos
-                        || excavationState == ExcavationState.CannotOutputMineDrop)) {
-                        targetOrePos = findOrePos();
-                    }
-
-                    if (excavationState == ExcavationState.NoOresInCurrentDepth
-                        || excavationState == ExcavationState.NotEnoughDrillTube) {
-                        tryDrillDownTube();
-                    } else if (targetOrePos != null && (excavationState == ExcavationState.InProgress
-                        || excavationState == ExcavationState.CannotOutputMineDrop)) {
-                        tryMine(targetOrePos, !getMineAll());
-                    }
-                }
+        if (miningEnergySpend >= getEnergyPerExcavation()) {
+            performNextCompletedStep();
+        }
 
-                if (excavationState == ExcavationState.InProgress) {
-                    miningSpentedEnergy -= getEnergyPerExcavation();
-                }
+        if (excavationState == InProgress && currentTickTime % 20 == 0) {
+            RecipeCrafter.soundHandler.playSound(false, this);
+        }
+
+    }
+
+    private void performNextCompletedStep() {
```

Comment:
performNextCompletedStep ??? Method is erg onduidelijk
Misschien idee om state machine toe te voegen?